src/Implementation/IExecutable.cs

```diff
@@ -4,6 +4,6 @@ namespace Symbolica.Implementation
 {
     public interface IExecutable
```

Comment:
Do you think we could rename this to `IExecutableState` now and make the method on it `ExecuteNextInstruction` again? I think that is more obvious from the outside so your counting stuff isn't just guessing what this thing is doing internallay and when it is yielding.

Reply:
I still don't think so. I'll break this into two questions though as I think they have slightly different answers.

> 1. "Should we name the method `ExecuteNextInstruction`. 

This almost certainly seems wrong to me, because broadly speaking it's the same as mixing up the difference between an `Iterable` and an `Iterator`. Calling it `ExecuteNextInstruction` would signify to me that I would be expected to keep calling this until some terminating condition was met, like you would with an `Iterator`. However, what we have here is now an `Iterable` and so what the client is really doing is saying, please run yourself until you hit a terminating condition and at each stage of the process return any new child states you've created.

To highlight this with an example, I think it would be odd to write:

```csharp
foreach (var forks in executable.ExecuteNextInstruction()) 
```

> 2. "Should we rename `IExecutable` to `IExecutableState`.

In light of my answer to the above I think not and I still think generally that we shouldn't need to leak the implementation details in the interface name.

On the broader point about making it more obvious that each element of the outer `IEnumerable` is actually the result of processing one instruction, I agree with you, but I think that's actually a downside of the way we've done it here. We're kind of artificially returning an `IEnumerable` just so that the client can count the number of elements that were in it. In doing so we've made it possible for the implementation of `IExecutable.Run` to potentially produce many forks before it finally terminates, but I don't think that is behaviour that we actually want or need. 

I'm struggling to see a way to convey the semantics that each element of the `IEnumerable` represents the processing of a single instruction, without either making the method name overly verbose or just adding this to the doc string for this method. Either way, I think it's an implicit part of the interface contract that in this implementation isn't going to be enforced by the type system.